File: LeetCode/Arrays/contains-duplicate.cc
/*
 * I did this in C++ beacuse I really didn't want to write a sorting algo in C
 * I did attempt to do a Range bucket in C with Shorts as the datatype over
 * the range of possible number in a signed int but leet code didn't want a 
 * huge array like that :P  I agree its not super space complexity efficient.
 *
 * https://leetcode.com/problems/contains-duplicate/
 * 
 * Runtime: 113 ms (62.93% Better)
 * Memory Usage: 46.7 MB (75.10% Better than)
 * 
 *  https://leetcode.com/PlaidCat/
 */

class Solution {
public:
    bool containsDuplicate(vector<int>& nums) {
      sort(nums.begin(), nums.end());
      for (int i =0; i < nums.size()-1; ++i){
        if (nums[i] == nums[i+1])
          return true;
      }
      return false;
    }
};

File: LeetCode/Arrays/contains-duplicate.rs
/*
 * Rustification of C++ code.
 * I'm finding it weird that I have to make a mutative copy to do this,
 * it might be just as easy to build a hash map rather than using the extra
 * memory right off the bat.
 *
 * https://leetcode.com/problems/contains-duplicate/
 * 
 * Runtime: 20 ms (53.85% Better)
 * Memory Usage: 3 MB (66.43% Better than)
 * 
 *  https://leetcode.com/PlaidCat/
 */

impl Solution {
  pub fn contains_duplicate(nums: Vec<i32>) -> bool {
    let mut my_nums: Vec<i32> = nums;
    my_nums.sort();
    for i in 0..(my_nums.len()-1) {
      if my_nums[i] == my_nums[i+1] {
        return true;
      }
    }
    return false;
  }
}